### PR TITLE
test: add test for mysql gatherGlobalVariables using sql-mock

### PR DESCRIPTION
**Pull request by fxedel**
### Required for all PRs:

- [x] ~Updated associated README.md.~
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

See #8561

Add a test that doesn't need a real database connection (by using sql-mock) and asserts proper parsing of the SQL query result.

I already found some issues with the version tag by adding this test, but I'd recommend adding the tests first and fixing the issues afterwards.

This is just PR #10036 re-opened, rebased to the master branch

Fixes: #8561

### CODE REVIEW

File: plugins/inputs/mysql/mysql_test.go [hunk 1]
```diff
@@ -178,6 +183,200 @@ func TestMysqlDNSAddTimeout(t *testing.T) {
...
+
+	for i, testCase := range testCases {
+		if testCase.name == "" {
+			testCase.name = fmt.Sprintf("#%d", i)
+		}
+
+		// uncomment this to run only a single test case:
+		// if testCase.name != "#1" {
+		// 	continue
+		// }
```
Comment (sspaink) [inline plugins/inputs/mysql/mysql_test.go:332]: Not sure if we want to keep this in the test? Seems like something you would only use when debugging these tests?
```suggestion
```

**Review by sspaink — APPROVED** [review]
Thank you for adding more tests! I think this looks good, looks like my suggested change caused gofmt to complain. I think you just need to run `make tidy` to fix the CI.

File: plugins/inputs/mysql/mysql_test.go [hunk 2]
```diff
@@ -178,6 +183,196 @@ func TestMysqlDNSAddTimeout(t *testing.T) {
...
+		// 		{"slave_type_conversions", "", "", false},
+		// 		{"tls_version", "TLSv1,TLSv1.1,TLSv1.2,TLSv1.3", "TLSv1,TLSv1.1,TLSv1.2,TLSv1.3", false},
+		// 		{"version", "8.0.27-0ubuntu0.20.04.1", "8.0.27-0ubuntu0.20.04.1", false},
+		// 		{"version_comment", "(Ubuntu)", "(Ubuntu)", false},
+		// 		{"version_compile_machine", "x86_64", "x86_64", false},
+		// 		{"version_compile_os", "Linux", "Linux", false},
+		// 		{"version_compile_zlib", "1.2.11", "1.2.11", false},
+		// 	},
+		// 	tags{"server": "127.0.0.1:3306", "version": "8.0.27-0ubuntu0.20.04.1"}, // TODO: Discuss which version tags should be used as tags
+		// },
```
Comment (reimda) [inline plugins/inputs/mysql/mysql_test.go:321]: What's your plan for these commented out tests? They look like valid problems that we should fix. If you don't want to fix them now, would you like to move them into another PR so we don't commit them commented out and forget about them?

**Review by reimda — CHANGES_REQUESTED** [review]
Could you also run `make fmt` to make CI pass? Thanks!

File: plugins/inputs/mysql/mysql_test.go (hunk 2 above)
Comment (fxedel) [reply plugins/inputs/mysql/mysql_test.go:321]: Hey @reimda, these are problems that we should fix, but I'm not sure yet what's the best way to fix. This first needs some discussion, so I'll open an issue and refer to this PR.

File: plugins/inputs/mysql/mysql_test.go [hunk 3]
```diff
@@ -178,6 +183,146 @@ func TestMysqlDNSAddTimeout(t *testing.T) {
...
+				}
+
+				if !assert.Truef(t, foundField, "unexpected field %s=%v (test case %q)", key, value, testCase.name) {
+					continue
+				}
+			}
+		}
+
+		for _, field := range testCase.fields {
+			assert.Truef(t, field.observed, "missing field %s=%v (test case %q)", field.key, field.parsedValue, testCase.name)
```
Comment (powersj) [inline plugins/inputs/mysql/mysql_test.go:321]: @fxedel,

Thanks for those updates and for taking the time to put this together! We had one final question regarding `assert` vs `require`. In all our other tests we try to prefer the use of `require` so we fail right away so it is very clear what is failing in tests.

I'm ready to +1 once this change is made.

Thanks!

**Review by powersj — APPROVED** [review]
Thank you for taking this on!
